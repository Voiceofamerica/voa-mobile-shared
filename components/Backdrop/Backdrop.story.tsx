
import 'react-native'
import * as React from 'react'
import * as RX from 'reactxp'
import { storiesOf, linkTo } from '@storybook/react-native'
import { action } from '@storybook/addon-actions'

const img = require('./Default.jpg')

import Backdrop from '.'

class ToggleBackdrop extends RX.Component<{ src?: any }, { blur: boolean }> {
  state = {
    blur: true,
  }

  toggleBlur = (blur = !this.state.blur) => {
    this.setState({ blur })
  }

  render () {
    return (
      <RX.View style={{ flex: 1, flexDirection: 'row' }}>
        <Backdrop imgSrc={this.props.src} defaultImgSrc={img} blur={this.state.blur} />
        <RX.Button onPress={() => this.toggleBlur()} style={{ padding: 10, height: 30, backgroundColor: '#DDD', margin: 20, borderRadius: 5 }}>
          <RX.Text>
            Toggle Blur
          </RX.Text>
        </RX.Button>
      </RX.View>
    )
  }
}

storiesOf('Backdrop', module)
  .add('display', () => (
    <Backdrop imgSrc={img} blur />
  ))
  .add('display unblurred', () => (
    <Backdrop imgSrc={img} />
  ))
  .add('togglable', () => (
    <ToggleBackdrop src={img} />
  ))
  .add('display with https resource', () => (
    <Backdrop blur imgSrc={'http://webneel.com/daily/sites/default/files/images/daily/09-2013/22-most-amazing-photos-sunrise.jpg'} />
  ))
  .add('togglable with http', () => (
    <ToggleBackdrop src={'http://webneel.com/daily/sites/default/files/images/daily/09-2013/22-most-amazing-photos-sunrise.jpg'} />
  ))
