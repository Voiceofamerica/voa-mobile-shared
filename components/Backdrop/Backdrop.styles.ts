
import * as RX from 'reactxp'

import { button, buttonText } from 'styles'

export default {
  container: RX.Styles.createViewStyle({
    position: 'absolute',
    top: 0,
    bottom: 0,
    left: 0,
    right: 0,
  }),
  image: RX.Styles.createImageStyle({
    width: '100%' as any,
    height: '100%' as any,
  }),
}
