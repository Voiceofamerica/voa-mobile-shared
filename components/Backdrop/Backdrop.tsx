
import * as React from 'react'
import { findNodeHandle } from 'react-native'
import { Image } from 'react-native'
import * as RX from 'reactxp'
import { BlurView } from 'react-native-blur'

import styles from './Backdrop.styles'

export interface Props {
  imgSrc: number | string
  defaultImgSrc?: number | string
  blur?: boolean
  blurDone?: (isBlurred: boolean) => void
}

export interface State {
  viewRef: any
  blurrinessVal: number
}

const MAX_BLUR = 10
const BLUR_CUTOFF = 1

class Backdrop extends RX.Component<Props, State> {
  state = {
    viewRef: null,
    blurrinessVal: 0,
  }

  blurrinessAnim = new RX.Animated.Value(0)
  blurId: string

  private backgroundImage: React.Component

  constructor (props, ctx) {
    super(props,ctx)
  }

  componentDidMount () {
    this.blurId = this.blurrinessAnim.addListener(this.blurListener)
  }

  componentWillUnmount () {
    this.blurrinessAnim.removeListener(this.blurId)
  }

  blurListener = ({ value: blurrinessVal }: any) => {
    const { blurDone = () => null } = this.props
    this.setState({ blurrinessVal })

    if (blurrinessVal === 1) {
      blurDone(true)
    } else if (blurrinessVal === 0) {
      blurDone(false)
    }
  }

  imageLoaded = () => {
    const viewRef = findNodeHandle(this.backgroundImage)
    this.setState({
      viewRef,
    })
    this.animateBlur(this.props.blur)
  }

  componentWillReceiveProps (nextProps: Props) {
    if (nextProps.blur !== this.props.blur) {
      this.animateBlur(nextProps.blur)
    }
  }

  animateBlur = (shouldBlur: boolean) => {
    const easing = RX.Animated.Easing.InOut()
    const toValue = shouldBlur ? 1 : 0

    RX.Animated.timing(
      this.blurrinessAnim,
      { toValue, duration: 500, easing },
    ).start()
  }

  renderBlur = () => {
    const { viewRef, blurrinessVal } = this.state

    const blurAmount = MAX_BLUR * blurrinessVal
    const opacity = (blurAmount - BLUR_CUTOFF) / (MAX_BLUR - BLUR_CUTOFF)

    if (blurAmount < BLUR_CUTOFF || !viewRef) {
      return null
    }

    return (
      <RX.View style={[styles.container, { opacity }]}>
        <BlurView
          style={styles.image as any}
          viewRef={viewRef}
          blurAmount={blurAmount}
          blurType='light' />
      </RX.View>
    )
  }

  renderImage = () => {
    const { imgSrc, defaultImgSrc = require('./Default.jpg') } = this.props

    const defaultSrc =
      typeof defaultImgSrc === 'number'
      ? defaultImgSrc
      : { uri: defaultImgSrc }

    const source =
      typeof imgSrc === 'number'
      ? imgSrc
      : { uri: imgSrc }

    return (
      <Image
        ref={img => this.backgroundImage = img}
        style={styles.image as any}
        source={source}
        defaultSource={defaultImgSrc && defaultSrc}
        onLoadEnd={this.imageLoaded}
        resizeMode='cover' />
    )
  }

  render () {
    return (
      <RX.View style={styles.container}>
        { this.renderImage() }
        { this.renderBlur() }
      </RX.View>
    )
  }
}

export default Backdrop
