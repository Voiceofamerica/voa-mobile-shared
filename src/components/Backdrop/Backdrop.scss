
.backdrop {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  right: 0;
  background-size: cover;
  background-position: center;
  transition: filter 1s;
  background: #444;

  .image {
    display: block;
    width: 100%;
    height: 100%;
    object-fit: cover;
  }
  
  .lightbox {
    position: fixed;
    top: 0;
    left: 0;
    bottom: 0;
    right: 0;
    opacity: 0;
    background: #FFF;
    transition: opacity 1s;
  }

  &.blurred {
    filter: blur(10px) grayscale(40%);

    .lightbox {
      opacity: 1;
    }
  }
}
