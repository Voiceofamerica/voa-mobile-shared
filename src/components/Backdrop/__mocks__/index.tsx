
import * as React from 'react'

export default (props) => (
  React.createElement('backdrop', { props }, props.children)
)
