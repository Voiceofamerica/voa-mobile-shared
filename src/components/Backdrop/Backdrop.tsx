
import * as React from 'react'
import { backdrop, blurred, image, lightbox } from './Backdrop.scss'

export interface Props {
  imgSrc: string
  blur?: boolean
}

const Backdrop = ({ blur = false, imgSrc }: Props) => {
  const cssClass = blur ? `${backdrop} ${blurred}` : backdrop
  return (
    <div className={cssClass}>
      <img src={imgSrc} className={image} />
      <div className={lightbox} />
    </div>
  )
}

export default Backdrop
