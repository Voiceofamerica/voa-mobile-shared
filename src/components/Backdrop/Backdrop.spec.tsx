
import * as React from 'react'
import { create } from 'react-test-renderer'
import Backdrop from './Backdrop'

describe('<Backdrop />', () => {
  describe('shapshots', () => {
    const sources = [
      'src1',
      'src2',
      'src3',
    ]

    sources.forEach(src => {
      it(`should render with the given imgSrc (${src})`, () => {
        const element = create((
          <Backdrop imgSrc={src} />
        ))

        expect(element).toMatchSnapshot()
      })
    })

    it('should blur if prop is set to true', () => {
      const element = create((
        <Backdrop imgSrc={'blurry'} blur />
      ))

      expect(element).toMatchSnapshot()
    })
  })
})
