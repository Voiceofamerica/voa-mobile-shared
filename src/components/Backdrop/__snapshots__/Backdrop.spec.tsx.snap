// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Backdrop /> shapshots should blur if prop is set to true 1`] = `
<div
  className="backdrop blurred"
>
  <img
    className="image"
    src="blurry"
  />
  <div
    className="lightbox"
  />
</div>
`;

exports[`<Backdrop /> shapshots should render with the given imgSrc (src1) 1`] = `
<div
  className="backdrop"
>
  <img
    className="image"
    src="src1"
  />
  <div
    className="lightbox"
  />
</div>
`;

exports[`<Backdrop /> shapshots should render with the given imgSrc (src2) 1`] = `
<div
  className="backdrop"
>
  <img
    className="image"
    src="src2"
  />
  <div
    className="lightbox"
  />
</div>
`;

exports[`<Backdrop /> shapshots should render with the given imgSrc (src3) 1`] = `
<div
  className="backdrop"
>
  <img
    className="image"
    src="src3"
  />
  <div
    className="lightbox"
  />
</div>
`;
